### DISCUSSION TITLE
How to add a number to photo

### DISCUSSION THREAD

**Discussion by satimis** [ImageMagick/ImageMagick#5899] [Help]
### ImageMagick version

8:6.9.11

### Operating system, version and so on

Ubuntu 22.04 desktop

### Description

I want to add a number, which changes, on the left top corner of all photos

For examples:-
photo-01.jpg, photo-02.jpg, photo-03.jpg, photo-04.jpg, photo-05.jpg, etc

I need to add a number on the left top corner of each photo
001, 002, 003, 004, 005 etc

Can ImageMagick do the job?

If YES, please advise HOW ?

Thanks in advance.

Regards

**Comment by fmw42** [comment]
Assuming you have IM 7, then

`magick image_01.jpg -gravity northwest -font arial -pointsize 24 -fill white -annotate +10+10 "001" image_01_annotate.jpg`

See https://imagemagick.org/Usage/text/#annotate

**Reply by satimis** [reply]
Hi fmw42,

Thanks for your advice

Can I do it on batch?  Instead of do it one-by-one ?

Thanks

Regards

**Reply by fmw42** [reply]
Yes, you can do it in a script FOR loop where you can control the numbering of the text.  Please always identify your IM version and platform/OS when asking questions since the syntax may vary.

**Reply by satimis** [reply]
Hi,

ImageMagick version;
$ apt policy imagemagick
```
imagemagick:
  Installed: 8:6.9.11.60+dfsg-1.3build2
  Candidate: 8:6.9.11.60+dfsg-1.3build2
  Version table:
 *** 8:6.9.11.60+dfsg-1.3build2 500
        500 http://hk.archive.ubuntu.com/ubuntu jammy/universe amd64 Packages
        100 /var/lib/dpkg/status
```

OS:
Ubuntu 22.04 desktop

Please guide me how to write the script.  I'm not very experienced in script writing.

Thanks in advance.

Regards

**Reply by fmw42** [reply]
How are you relating the image with the text number?

**Reply by satimis** [reply]
Starting;
001, 002, 003, 004, 005 ...... etc
added on the left top corner

**Reply by fmw42** [reply]
You do not understand the question.  Are you taking every image in a directory?  Are they numbered the way you want in your images so that the first image will get 001 as the text no matter what the image file name is?  Are your images numbered in the directory to correspond to your number.  Image_01.jpg will get 001. What happens to your list of images at image_99.jpg when the next one is image_100.jpg. (You list images as image_01.jpg and numbers as 001.  This is not a good correspondence if you have more than 99 images). So Is image_100.jpg in the directory list where you need it after image_99.jpg or is it out of place.  

It is easy to just take the files in order as listed in the directory and assign them a running number increasing from  001. But then if the images are number and not in increasing order, then the text will not correspond.  

**Please confirm that all your images are in sequentially increasing order in your directory.**

**Reply by satimis** [reply]
`Please confirm that all your images are in sequentially increasing order in your directory.`
Thanks for your advice.
Now I understand all images must be in sequentially increasing order in the directory otherwise the script won't work?
I must add the text to the image one-by-one ?

Pls advise;
-gravity northwest (the text added on the left top corner)
Whether;
-gravity southwest (the text added at the left bottom corner) ?
-gravity northeast (the text added on the right top corner) ?
-gravity southeast (the text added at the right bottom corner) ?

How about;
Text on the top center ?
Text at the bottom center ?
Text on the left center and rotated vertically ?
Text on the right center and rotated vertically ?

Thanks
Regards

**Reply by satimis** [reply]
**Edit:**
It is very strange;

Command 'magick' not found on Ubuntu 22.04

$ convert -version
```
Version: ImageMagick 6.9.11-60 Q16 x86_64 2021-01-25 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP(4.5) 
Delegates (built-in): bzlib djvu fftw fontconfig freetype heic jbig jng jp2 jpeg lcms lqr ltdl lzma openexr pangocairo png tiff webp wmf x xml zlib
```

**Reply by satimis** [reply]
Tried again, replacing "magick' with "convert"

$ convert image_01.jpg -gravity northwest -font arial -pointsize 90 -fill red -annotate +100+100 "001" image_01_annotate.jpg 

This time it works

See attached file
![screenshot_convert30](https://user-images.githubusercontent.com/30752517/209122787-df7c192e-293c-424d-83d1-b8c7ea087cd6.png)

How to add "Bold" ?

**Reply by fmw42** [reply]
Top center is north, bottom center is south. Right side center is east. Left side center is west.

There is not bold. You either make it larger or get a font that is a bold font.

**Reply by fmw42** [reply]
This will do a FOR loop over all JPG images in a directory and add your numbered label. It takes the images sequentially as listed in your directory and adds the number in the corner sequentially from 001. It writes the output to the same directory as the input.  If you want a different directory, the create an empty folder and put the path as part of the output. Give it a try and let us know.  I suggest you put just a few images in a new directory to test it.

```
cd
cd path_to_your_image_directory
i=1
for img in *.jpg; do
name=$(convert "$img" -format "%t" info:)
ii=`printf "%03d" "$i"`
convert "$img" -gravity northwest -font arial -pointsize 90 -fill red -annotate +100+100 "$ii" ${name}_annotate.jpg
i=$((i+1))
done

```

**Reply by satimis** [reply]
Hi Fred,

Thanks for your advice.  It works seamlessly.

Performed following steps;
1. Create a new folder for this test
2. Copy 10 images on this folder
3. On Terminal (Ubuntu 22.04 desktop)

Run
```
$ cd folder_for_test/
$ i=1
$ for img in *.jpg; do
> name=$(convert "$img" -format "%t" info:)
> convert "$img" -gravity northwest -font arial -pointsize 90 -fill red -annotate +100+100 "$ii" ${name}_annotate.jpg
> i=$((i+1))
> done
```

The number is added to the left top corner on each image.  

1.  Could you please explain your command?
2.  How to create your script as a file for execution?

A further question.
An image in the folder is of smaller size.  Then the number added becomes too large.
(pls see attached image)

Any solution?  Thanks in advance.

Regards
![screenshot_large_number80](https://user-images.githubusercontent.com/30752517/209276162-471f01bc-108c-4727-ae55-53dc460a0007.png)

**Reply by satimis** [reply]
Hi Fred,
`
Right side center is east. Left side center is west.`
The number is displayed horizontally.  How to rotate it 90 deg ?

How to find out whether there are Italic and Bold fonts installed on the OS ?

Thanks

Regards

**Reply by fmw42** [reply]
```
cd
cd path_to_your_image_directory
i=1
for img in *.jpg; do
name=$(convert "$img" -format "%t" info:)
ii=`printf "%03d" "$i"`
convert "$img" -gravity northwest -font arial -pointsize 90 -fill red -annotate +100+100 "$ii" ${name}_annotate.jpg
i=$((i+1))
done
```

```
Line1: set diretory to home directory
Line2: change directory to your image folder
Line3: Set counter to 1
Line4: For loop over all JPG files in the folder (*.jpg means all filenames that have jpg suffix)
Line5: find the name of the image without its suffix
Line5: convert the count to have leading zeros
Line6: do your convert command to add the label
Line7: increment the counter
Line8: end the For loop

```

You can rotate the label using the XxY on the -annotate. See https://imagemagick.org/script/command-line-options.php#annotate and https://imagemagick.org/Usage/text/#annotate.

Or rotate the image, add the text, then rotate the image back again.

**Reply by satimis** [reply]
Hi Fred,

Lot of thanks for your explanation.  Please clarify follows;

1. Line3: Set counter to 1
Whether i=increment of the counter ?

2. Please explain in more detail on following line;
name=$(convert "$img" -format "%t" info:)

3. Line5: convert the count to have leading zeros
ii=`printf "%03d" "$i"`
What is ii ?

Thanks

Furthermore;
1. If I need adding a fixed prefix to the number;
e.g. AA001
AA won't change and in YELLOW color
What will be the command ?

2. If I only need 2 digits on the number
e.g. AA01
What will be the command ?

Thanks

Regards.

**Reply by satimis** [reply]
Hi Fred,

Still I couldn't find a solution to rotate the text 001 90deg
Whether I need to create the text as image first?
Please help. Thanks

Regards

**Reply by fmw42** [reply]
1) i is the counter. I set it to 1 to begin and increment it at the end

2) name=$(convert "$img" -format "%t" info:)   sets the variable name to the part of the file before the .suffix.  %t is a string format for that.  See https://imagemagick.org/script/escape.php

3) ii=$(printf "%03d" "$I")   ii is a new variable that I named to hold the form of the counter that has leading zeros as 001.  To have  leading zeros as 01 use %02d.

I do not understand your question about YELLOW color. But if you want yellow text, then change -fill black to -fill yellow,

To have AA001, use "AA$ii" in place of "$ii"

To rotate text

convert lena.png -fill black -gravity center -pointsize 36 -font ubuntu -annotate 45x45+0+0 "This is text" x.png

![x](https://user-images.githubusercontent.com/13458218/209422693-911f41b8-c90e-4107-89bc-a6a50bed5a13.png)

the 45x45+0+0 means 45x45 degrees and +0+0 part is the offset from the center

**Reply by satimis** [reply]
Hi Fred,
`I do not understand your question about YELLOW color. But if you want yellow text, then change -fill black to -fill yellow,`

On the number: AA001
AA not changing and in YELLOW color
001 changing but in RED color

To change the number to AA01
Please advise how to change your command line ?

Thanks
Regards

**Reply by fmw42** [reply]
You cannot have two colors for the same text string in one -annotate command. You could do it with two successive -annotates, but you would have to compute the offset. That would be easier with a fixed width font where each character takes the same space. Or you could do it with PANGO. But that is different. See the page at https://imagemagick.org/Usage/text/#pango

**Reply by satimis** [reply]
Hi Fred,
`Or you could do it with PANGO. But that is different. See the page at https://imagemagick.org/Usage/text/#pango`
Noted and thanks

Regards

**Reply by satimis** [reply]
Hi Fred,
`To have AA001, use "AA$ii" in place of "$ii"`
Your advice works for me.   Thanks
If I only need AA01 what shall I change?

Regards

**Reply by fmw42** [reply]
As I said above:

> 3) ii=$(printf "%03d" "$I")   ii is a new variable that I named to hold the form of the counter that has leading zeros as 001.  To have  leading zeros as 01 use %02d.

So use %02d in place of %03d

**Reply by satimis** [reply]
`So use %02d in place of %03d`
It works.  Thanks
Regards
